### PR TITLE
Initial architecture jaxb

### PR DESCRIPTION
**Pull request by piumika**
Hello @tonivdv ,
I did some research to see if we could use xjc to generate the java domain classes automatically using the dtd (something similar to what we did in smart).But the oas dtd is not in correct format therefore we cannot use it to automatically generate the java classes.

However , we can use jaxb along with dom4j and use jaxb to generate objects from the xml and also generate the xml from the objects.In this branch I have used jaxb in the AdvertiserService and would like your opinion on using jaxb.Please let me know your option so that we can use jaxb for the project or parse the xml the way we did before.(Parsing the xml the old way might be faster as we already have the code) 

Also let us know what you think of the structure/architecture of the project. Is there anything that is inappropriate ?

cc @gunith

### CODE REVIEW

**Comment by tonivdv** [comment]
Hello @piumika ,

Whether we parse it manually or with a library is for me that important. What I do feel is important is to have the entity classes (domain class). Always try to think how a developer is going to use this library.

So I think starting with the documentation directly would be a good thing too, as it will show you if it's easy to use or not.

As a general remark, I would not use Adl\* names in the classes, and think about adding documentation as it's not always easy to understand why or what a class if for.

File: src/main/java/eu/adlogix/appnexus/oas/client/ExceptionUtils.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package eu.adlogix.appnexus.oas.client;
+
+public final class ExceptionUtils {
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/ExceptionUtils.java:3]: What's the advantage of having this class?

File: src/main/java/eu/adlogix/appnexus/oas/client/certificate/CertificateManager.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+	private static final String CONF_DIRECTORY = "conf";
+	private static final String XAXIS_KEYSTORE_FILENAME = "xaxisCACerts";
+
+	private static final String HTTPS_PREFIX = "https://";
+
+	public synchronized void prepareKeyStoreForHost(final String host) {
+
+		final File confDir = new File("conf");
+
+		if (!confDir.exists())
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/certificate/CertificateManager.java:45]: can be written in one line

File: src/main/java/eu/adlogix/appnexus/oas/client/certificate/CertificateManager.java [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+			if (!usedKeyStoreFile.delete())
+				throw new RuntimeException("Unable to delete pre-existing keyStore at "
+						+ usedKeyStoreFile.getAbsolutePath());
+
+			if (usedKeyStoreFile.exists())
+				throw new RuntimeException("We could not delete the existing keyStore at "
+						+ usedKeyStoreFile.getAbsolutePath());
+		}
+
+		if (!host.startsWith(HTTPS_PREFIX))
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/certificate/CertificateManager.java:63]: this verification should be done sooner no?

File: src/main/java/eu/adlogix/appnexus/oas/client/domain/Advertiser.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	String organization;
+	@XmlElement(name = "BillingInformation")
+	BillingInformation billingInformation;
+	@XmlElement(name = "InternalQuickReport")
+	String internalQuickReport;
+	@XmlElement(name = "ExternalQuickReport")
+	String externalQuickReport;
+
+	public Advertiser(String id, String organization) {
+		this();
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/domain/Advertiser.java:32]: I would do it the other way around. The constructor with no args, calling the construct with args.

File: src/main/java/eu/adlogix/appnexus/oas/client/domain/Advertiser.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import lombok.Data;
+
+@XmlType(propOrder = { "id", "organization", "billingInformation", "internalQuickReport", "externalQuickReport" })
+@XmlRootElement(name = "Advertiser")
+@XmlAccessorType(XmlAccessType.FIELD)
+@Data
+public class Advertiser {
+
+	@XmlElement(name = "Id")
+	String id;
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/domain/Advertiser.java:21]: why package level? And not simply private?

File: src/main/java/eu/adlogix/appnexus/oas/client/domain/BillingInformation.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import lombok.Data;
+
+@Data
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlType(propOrder = { "method", "country" })
+@XmlRootElement(name = "BillingInformation")
+public class BillingInformation {
+
+    @XmlElement(name = "Method")
+    protected Method method;
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/domain/BillingInformation.java:18]: Why protected level and not private?

File: src/main/java/eu/adlogix/appnexus/oas/client/domain/Code.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlType(propOrder = { "value" })
+@XmlRootElement(name = "Code")
+public class Code {
+
+    @XmlValue
+    protected String value;
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/domain/Code.java:22]: Why protected level and not private?

File: src/main/java/eu/adlogix/appnexus/oas/client/domain/Country.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlType(propOrder = { "code" })
+@XmlRootElement(name = "Country")
+public class Country {
+
+    @XmlElement(name = "Code")
+    protected List<Code> code;
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/domain/Country.java:24]: Why protected level and not private?

File: src/main/java/eu/adlogix/appnexus/oas/client/domain/Method.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+@Getter
+@AllArgsConstructor
+@NoArgsConstructor
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlType(propOrder = { "code" })
+@XmlRootElement(name = "Method")
+public class Method {
+
+    @XmlElement(name = "Code", required = true)
+    protected Code code;
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/domain/Method.java:22]: Why protected level and not private?

File: src/main/java/eu/adlogix/appnexus/oas/client/service/AbstractOasService.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+import eu.adlogix.appnexus.oas.client.domain.PushLevel;
+import eu.adlogix.appnexus.oas.client.util.Credentials;
+import eu.adlogix.appnexus.oas.client.xml.ResponseParser;
+import eu.adlogix.appnexus.oas.client.xml.XmlRequestGenerator;
+import eu.adlogix.appnexus.oas.client.xml.ResponseParser.ResponseElementHandler;
+import eu.adlogix.appnexus.oas.client.xml.ResponseParser.ResponseObjectHandler;
+import eu.adlogix.appnexus.oas.utils.log.LogUtils;
+
+
+public class AbstractOasService {
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/service/AbstractOasService.java:19]: It's call Abstract, but it's not an abstract class? I would also introduce a OasService Interface that this class implementats

File: src/main/java/eu/adlogix/appnexus/oas/client/service/AdvertiserService.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+public class AdvertiserService extends AbstractOasService {
+
+	private final XmlRequestGenerator addAdvertiserRequestGenerator = new XmlRequestGenerator("add-advertiser");
+	private final XmlRequestGenerator listAdvertisersRequestGenerator = new XmlRequestGenerator("list-advertisers");
+	private final XmlRequestGenerator readAdvertisersRequestGenerator = new XmlRequestGenerator("read-advertiser");
+	private final XmlRequestGenerator updateAdvertiserRequestGenerator = new XmlRequestGenerator("update-advertiser");
+
+	private XmlGenerator xmlGenerator = new XmlGenerator();
+
+	public AdvertiserService(Properties credentials) {
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/service/AdvertiserService.java:26]: It's a bit annoying that for each Service we will have to pass the credentials.

File: src/main/java/eu/adlogix/appnexus/oas/client/service/OasApiService.java [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+			try {
+				res = (String) call.invoke(new Object[] { account, user, password, adXML });
+			} catch (Exception e) {
+				if (retryOnConnectionErrors) {
+					logger.error("Exception while calling OAS API: [" + e.toString() + "]", e);
+
+					retryCount++;
+
+					if (retryCount < MAX_RETRY) {
+						logger.info("Xaxis API call retry #" + retryCount);
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/service/OasApiService.java:92]: Xaxis should be OAS

File: src/main/java/eu/adlogix/appnexus/oas/client/service/OasApiService.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+					} else {
+						logger.error("We just exceeded the maximal number of Xaxis API calls retries while querying ["
+								+ adXML + "]");
+						throw new RuntimeException("Too many Xaxis API calls retries", e);
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/service/OasApiService.java:96]: We should have a better error naming. Not sure if it's clear.

File: src/main/java/eu/adlogix/appnexus/oas/client/util/Credentials.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+import org.apache.commons.lang3.StringUtils;
+
+import eu.adlogix.appnexus.oas.client.domain.PushLevel;
+
+
+public final class Credentials {
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/util/Credentials.java:16]: Not sure though, but you should check 'spring boot'. I think they have clean ways for external parameters and files

File: src/main/java/eu/adlogix/appnexus/oas/client/utils/AdlFileUtils.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URL;
+
+import org.apache.commons.io.IOUtils;
+
+import eu.adlogix.appnexus.oas.utils.file.AdlResourceNotFoundException;
+
+public class AdlFileUtils {
```
Comment (tonivdv) [inline src/main/java/eu/adlogix/appnexus/oas/client/utils/AdlFileUtils.java:15]: Can't this one be mixed with ResourceUtils?

**Comment by piumika** [comment]
Hello @tonivdv ,

Thanks for your comments. Will do the request/response parsing manually (without using jaxb) as we already have the code and will be faster. Will incorporate the improvements you mentioned and document the classes.
Thanks

@gunith

**Comment by tonivdv** [comment]
@piumika just as a note. Your choice should not be done on 'fast' but on the best solution. So be sure that doing it manually is the best approach!

**Comment by piumika** [comment]
Hello @tonivdv ,

Both methods have pros and cons. The jaxb solution is cleaner as we do not have to deal with manual parsing and its annotation based.

In certain sections of the network service and report service manual parsing seems to be a better approach than jaxb as we will have more control of the parsing and the attributes of the domain objects.

So after discussing we came to the conclusion to use manual parsing as it will be consistent.
cc @gunith

**Pull request closed without merging**